File: medium/3.longest-substring-without-repeating-characters.py
"""
abcab*cbb^


pw*wke^w
"""
#abc

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        left = 0
        right = 0
        dist_max = 0
        while right < len(s):
            char = s[right]
            if char in seen:
                seen.remove(s[left])
                left += 1
            else:
                seen.add(char)
                right += 1
            #print(seen)
            dist_max = max(right - left, dist_max)
        return dist_max
            

File: medium/3.longest-substring-without-repeating-characters.rs
use std::cmp::max;
use std::collections::HashSet;
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let (mut l, mut max_len) = (0, 0);
        let mut hs = HashSet::new();
        let chars:Vec<_> = s.chars().collect();
        for (i, c) in chars.iter().enumerate() {
            while hs.contains(&c) {
                hs.remove(&chars[l]);
                l += 1;
            }
            hs.insert(c);
            max_len = max(max_len, hs.len());
        }
        max_len as i32
    }
}